### PR TITLE
add transformation matrix

### PR DESCRIPTION
**Pull request by Licini**
Adding transform to base object:

- added translation, rotation (Euler XYZ), scale vector
- matrix dynamically updated from above vectors

Example:
`scripts/v120_transform.py`

image

### CODE REVIEW

File: src/compas_view2/objects/object.py [hunk 1]
```diff
@@ -34,3 +64,50 @@ def create(self):
...
+        return self._transformation.matrix
+
+    @matrix.setter
+    def matrix(self, matrix):
+        """Set the object's translation, rotation and scale from given matrix, and update object's matrix"""
+        scale, _, rotation, tranlation, _ = decompose_matrix(matrix)
+        self.translation = tranlation
+        self.rotation = rotation
+        self.scale = scale
+        self._update_matrix()
```
Comment (brgcode) [inline src/compas_view2/objects/object.py:113]: why first decompose and then again compose? to exclude the projection?

File: src/compas_view2/objects/object.py [hunk 2]
```diff
@@ -34,3 +64,50 @@ def create(self):
...
+    def rotation(self, data):
+        self._rotation = data
+
+    @property
+    def scale(self):
+        return self._scale
+
+    @scale.setter
+    def scale(self, data):
+        self._scale = data
```
Comment (brgcode) [inline src/compas_view2/objects/object.py:90]: why use setters and getters if we don't process the input?

File: src/compas_view2/objects/object.py [hunk 3]
```diff
@@ -34,3 +64,50 @@ def create(self):
...
+
+    def _update_matrix(self):
+        """Update the matrix from object's translation, rotation and scale"""
+        T1 = Translation.from_vector(self.translation)
+        R1 = Rotation.from_euler_angles(self.rotation)
+        S1 = Scale.from_factors(self.scale)
+        M = T1 * R1 * S1
+        self._transformation.data = M.data
```
Comment (brgcode) [inline src/compas_view2/objects/object.py:98]: not sure if this is the most efficient way to go about this

File: src/compas_view2/shaders/120/mesh.vert [hunk 4]
```diff
@@ -5,6 +5,7 @@ attribute vec3 color;
 
 uniform mat4 projection;
 uniform mat4 viewworld;
+uniform mat4 transform;
```
Comment (brgcode) [inline src/compas_view2/shaders/120/mesh.vert:8]: is transform not a reserved keyword in some flavours of glsl?

**Review by brgcode — CHANGES_REQUESTED** [review]

File: src/compas_view2/objects/object.py (hunk 1 above)
Comment (Licini) [reply src/compas_view2/objects/object.py:113]: The is to automatically update the translation, rotation and scale while user directly change the matrix. Decomposing and composing ensures those three vectors and matrix are always consistent. 

 If we are to assign the matrix directly then we would need to include shear and projection as object attribute too

File: src/compas_view2/objects/object.py (hunk 2 above)
Comment (Licini) [reply src/compas_view2/objects/object.py:90]: A mistake of mine.... was intended to do 
`self._translation.data = data`
updating now

File: src/compas_view2/shaders/120/mesh.vert (hunk 4 above)
Comment (Licini) [reply src/compas_view2/shaders/120/mesh.vert:8]: Says that all glsl reserved names start with `gl_`

https://www.khronos.org/opengl/wiki/Core_Language_(GLSL)#:~:text=to%20be%20compatibility%20.-,Reserved%20names,as%20identifiers%20in%20any%20context.

File: src/compas_view2/objects/object.py (hunk 3 above)
Comment (Licini) [reply src/compas_view2/objects/object.py:98]: You mean the method to compose the matrix or the general behavior of keep updating the matrix?

Comment (brgcode) [reply src/compas_view2/objects/object.py:98]: both :) but not important now...

the data thing seems wrong though. why not just do `self._transformation = M` or `self._transformation.matrix = M.matrix`?

although also the latter seems redundant...

File: src/compas_view2/shaders/120/mesh.vert (hunk 4 above)
Comment (brgcode) [reply src/compas_view2/shaders/120/mesh.vert:8]: ok

File: src/compas_view2/objects/object.py (hunk 2 above)
Comment (brgcode) [reply src/compas_view2/objects/object.py:90]: don't think that makes sense. just assign the object, no? or if necessary the matrix, but that also seems redundant...

File: src/compas_view2/objects/object.py [hunk 5]
```diff
@@ -34,3 +64,50 @@ def create(self):
...
+    def translation(self, data):
+        self._translation.data = data
+
+    @property
+    def rotation(self):
+        return self._rotation
+
+    @rotation.setter
+    def rotation(self, data):
+        self._rotation.data = data
```
Comment (brgcode) [inline src/compas_view2/objects/object.py:82]: i don't understand how this works with the `_update_matrix` function.
what is `self._rotation`? the update function seems to suggest that this is a triplet of angles, but here it suggests that it is a compas (transformation?) object.
same for the translation and scale.

File: src/compas_view2/objects/object.py [hunk 6]
```diff
@@ -34,3 +64,50 @@ def create(self):
...
     def edit(self):
         pass
+
+    @property
+    def translation(self):
+        return self._translation
+
+    @translation.setter
+    def translation(self, data):
+        self._translation.data = data
```
Comment (brgcode) [inline src/compas_view2/objects/object.py:74]: should this not be
```python
@translation.setter
def translation(self, vector):
    self._translation = Vector(*vector)
```

File: src/compas_view2/objects/object.py (hunk 3 above)
Comment (Licini) [reply src/compas_view2/objects/object.py:98]: yep was looking at the code for matrix...... I wanted to make sure it is only data copied without replacing the entire object. And realised the data setter does that anyway....

File: src/compas_view2/objects/object.py (hunk 5 above)
Comment (brgcode) [reply src/compas_view2/objects/object.py:82]: should it not be something like
```
@rotation.setter
def rotation(self, angles):
    assert len(angles) == 3  # not sure if this is necessary and/or makes sense
    self._rotation = angles

File: src/compas_view2/objects/object.py (hunk 6 above)
Comment (brgcode) [reply src/compas_view2/objects/object.py:74]:
```python
@translation.setter
def translation(self, vector):
    self._translation[0] = vector[0]
    self._translation[1] = vector[1]
    self._translation[2] = vector[2]
```

File: src/compas_view2/objects/object.py [hunk 7]
```diff
@@ -34,3 +63,56 @@ def create(self):
...
+        self._translation[0] = vector[0]
+        self._translation[1] = vector[1]
+        self._translation[2] = vector[2]
+
+    @property
+    def rotation(self):
+        return self._rotation
+
+    @rotation.setter
+    def rotation(self, rotation_euler):
```
Comment (brgcode) [inline src/compas_view2/objects/object.py:82]: this is just personal preference but i would write `def rotation(self, angles)`

File: src/compas_view2/objects/object.py [hunk 8]
```diff
@@ -34,3 +63,56 @@ def create(self):
...
+        self._rotation[0] = rotation_euler[0]
+        self._rotation[1] = rotation_euler[1]
+        self._rotation[2] = rotation_euler[2]
+
+    @property
+    def scale(self):
+        return self._scale
+
+    @scale.setter
+    def scale(self, scale_factor):
```
Comment (brgcode) [inline src/compas_view2/objects/object.py:92]: also just personal preference `def scale(self, factors)`

**Review by brgcode — APPROVED** [review]

**Comment by brgcode** [comment]
just two small things. everything else looks good!

**Comment by Licini** [comment]
Done @brgcode

**Pull request merged**